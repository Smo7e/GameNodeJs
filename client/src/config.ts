// UdSU
// const DOMAIN = 'http://studfront';
// const PORT = 81;

// dev
const DOMAIN = "http://localhost";
const PORT = 3001;

// prod
//...

export const HOST = PORT ? `${DOMAIN}:${PORT}` : `${DOMAIN}`;

export const MEDIATOR = {
    EVENTS: {
        SERVER_ERROR: "SERVER_ERROR",
        GET_MESSAGES: "GET_MESSAGES",
        GET_SCENE: "GET_SCENE",
        GET_USER: "GET_USER",
        GET_GAMERS: "GET_GAMERS",
        GET_FRIENDS: "GET_FRIENDS",
        GET_INVITES: "GET_INVITES",
        LOGIN: "LOGIN",
        SIGNUP: "SIGNUP",
        LOGOUT: "LOGOUT",
        GET_MOBS: "GET_MOBS",
        GET_ERROR: "GET_ERROR",
        PLAY_MUSIC: "PLAY_MUSIC", 
        STOP_MUSIC: "STOP_MUSIC", 
        SET_MUSIC_VOLUME: "SET_MUSIC_VOLUME", 
        UPDATE_ARR_BULLET_TRAJECTORY: "UPDATE_ARR_BULLET_TRAJECTORY",
    },
    TRIGGERS: {},
};
