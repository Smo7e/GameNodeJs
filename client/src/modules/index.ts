import { TError } from "./Server/types";
import Server from "./Server/Server";
import Mediator from "./Mediator/Mediator";
import Store from "./Store/Store";

export type { TError };
export { Server, Mediator, Store };
