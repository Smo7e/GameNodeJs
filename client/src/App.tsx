import React, { useState } from "react";
import { HOST, MEDIATOR } from "./config";
import { Server, Mediator, Store, TError } from "./modules";
import SignUp from "./component/SignUp/SignUp";
import Game from "./modules/Game/Game";
import Menu from "./component/Menu/Menu";
import Login from "./component/Login/Login";
import Heroes from "./component/Heroes/Heroes";
import Interface from "./component/Interface/Interface";
import Lobby from "./component/Lobby/Lobby";
import ErrorMessage from "./modules/ErrorMessage/ErrorMessage";
import Parameters from "./component/Parameters/Parameters";
import Music from "./component/Parameters/Music";
import "./App.css";

export const ServerContext = React.createContext<Server>(null!);
export const MediatorContext = React.createContext<Mediator>(null!);
export const StoreContext = React.createContext<Store>(null!);

export enum EPAGES {
    SIGNUP,
    LOGIN,
    GAME,
    MENU,
    HEROES,
    LOBBY,
    PARAMETERS,
}

const MainApp = () => {
    const [epages, setEpages] = useState<EPAGES>(EPAGES.LOGIN);
    return (
        <>
            {epages === EPAGES.LOGIN ? (
                <Login epages={setEpages} />
            ) : epages === EPAGES.SIGNUP ? (
                <SignUp epages={setEpages} />
            ) : epages === EPAGES.MENU ? (
                <Menu epages={setEpages} />
            ) : epages === EPAGES.GAME ? (
                <>
                    <Game />
                    <Interface />
                </>
            ) : epages === EPAGES.HEROES ? (
                <Heroes epages={setEpages} />
            ) : epages === EPAGES.LOBBY ? (
                <Lobby epages={setEpages} />
            ) : epages === EPAGES.PARAMETERS ? (
                <Parameters epages={setEpages} />
            ) : (
                <></>
            )}
            <ErrorMessage />
            <Music />
        </>
    );
};

const App: React.FC = () => {
    const store = new Store();
    const mediator = new Mediator(MEDIATOR);
    const server = new Server(HOST, mediator, store);

    const { SERVER_ERROR } = mediator.getEventTypes();
    mediator.subscribe(SERVER_ERROR, (data: TError) => {
        console.log(data);
    });

    return (
        <StoreContext.Provider value={store}>
            <MediatorContext.Provider value={mediator}>
                <ServerContext.Provider value={server}>
                    <MainApp />
                </ServerContext.Provider>
            </MediatorContext.Provider>
        </StoreContext.Provider>
    );
};

export default App;
