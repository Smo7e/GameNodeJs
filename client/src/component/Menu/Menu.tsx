import React, { useContext, useEffect, useRef, useState } from "react";
import { EPAGES, MediatorContext, ServerContext } from "../../App";
import logo from "./image/logo.png";
import "./Menu.css";
import { TFriend, TGamer, TInvites } from "../../modules/Server/types";
interface IMenuProps {
    epages: Function;
}
const Menu: React.FC<IMenuProps> = ({ epages }) => {
    const server = useContext(ServerContext);
    const mediator = useContext(MediatorContext);

    const idRef = useRef<HTMLInputElement>(null);
    const [friends, setFriends] = useState<TFriend[]>([{ id: 0, name: "У Вас нету друзей,пхахахахахха!" }]);
    const [invites, setInvites] = useState<TInvites | null>(null);
    const [showIdInput, setShowIdInput] = useState(false);

    const clickHandler = () => {
        const id = idRef.current!.value;
        server.addFriend(id);
    };
    const toggleIdInput = () => {
        setShowIdInput(!showIdInput);
    };

    useEffect(() => {
        const { GET_INVITES } = mediator.getEventTypes();
        const { GET_FRIENDS } = mediator.getEventTypes();
        const { LOGOUT } = mediator.getEventTypes();

        const getInvitesHandler = (invites: TInvites) => {
            setInvites(invites);
        };
        const getFriendsHandler = (friends: TFriend[]) => {
            setFriends(friends);
        };
        const logoutHandler = () => {
            epages(EPAGES.LOGIN);
        };
        mediator.subscribe(GET_FRIENDS, getFriendsHandler);
        mediator.subscribe(GET_INVITES, getInvitesHandler);
        mediator.subscribe(LOGOUT, logoutHandler);

        return () => {
            mediator.unsubscribe(GET_FRIENDS, getFriendsHandler);
            mediator.unsubscribe(GET_INVITES, getInvitesHandler);
            mediator.unsubscribe(LOGOUT, logoutHandler);
        };
    });
    const lobbyHandler = async () => {
        await server.createLobby();
        epages(EPAGES.LOBBY);
    };

    // mp

    return (
        <div className="mainMenu" id="test-mainMemu">
            <div style={{ height: 200, width: 200, position: "absolute", bottom: 200, right: 200 }}></div>
            <img className="photo-button" src={logo} id="test-logo" />

            <div className="buttons-container">
                <div onClick={lobbyHandler} className="button1" id="test-play">
                    Играть
                </div>
                <div onClick={() => epages(EPAGES.HEROES)} className="button2" id="test-heroes">
                    Герои
                </div>
                <div onClick={() => epages(EPAGES.PARAMETERS)} className="button3" id="test-settings">
                    Параметры
                </div>
            </div>

            <div className="profile-panel" id="test-profile">
                <div className="user-profile" id="test-user"></div>
                <hr className="hr-user-profile1" id="test-hr1" />

                <div className="text-button" id="test-friends">
                    Друзья
                    {showIdInput && (
                        <div id="test-new-profile">
                            <input className="add-friend" ref={idRef} type="text" placeholder="Введите ID друга" />
                            <button className="add-friend-button" onClick={clickHandler}>
                                Добавить друга
                            </button>
                        </div>
                    )}
                </div>

                <div className="text-Invitation" id="test-friends">
                    Приглашение
                    <hr className="hr-line-Invitation" id="test-friends" />
                </div>

                <div onClick={toggleIdInput} className="new-profile-button"></div>
                <div
                    style={{
                        left: 5,
                        height: 240,
                        width: 222,
                        position: "absolute",
                        bottom: 60,
                        right: 200,
                        overflowY: "hidden",
                    }}
                >
                    {invites && invites.friendsId ? (
                        invites.friendsId.map((invite: number, index: number) => (
                            <div style={{ display: "flex" }} key={index}>
                                <div className="your-friend" key={index}>
                                    Пользователь с id:{invite} приглашает вас.
                                </div>
                                <button
                                    className="toaccept"
                                    onClick={() => {
                                        server.addGamers(invites!.lobbyName);
                                        epages(EPAGES.LOBBY);
                                    }}
                                ></button>
                                <button className="deny" onClick={() => console.log(0)}></button> ////////
                            </div>
                        ))
                    ) : (
                        <></>
                    )}
                </div>

                <div className="your-friend-menu" id="test-friend-menu">
                    {friends.map((friend, index) => (
                        <div className="your-friend" key={index}>
                            {`${friend.id}: ${friend.name}`}
                        </div>
                    ))}
                </div>
                <hr className="hr-user-profile2" id="test-hr2" />
            </div>
            <button className="button-account" id="test-change-account" onClick={() => server.logout()}>
                Сменить аккаунт
            </button>
        </div>
    );
};

export default Menu;
