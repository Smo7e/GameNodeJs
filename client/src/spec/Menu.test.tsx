import { render, screen } from "@testing-library/react";
import Menu from "../component/Menu/Menu";
import { MediatorContext, ServerContext } from "../App";
import { Mediator, Server, Store } from "../modules";
import { HOST, MEDIATOR } from "../config";

const store = new Store();
const mediator = new Mediator(MEDIATOR);
const server = new Server(HOST, mediator, store);

test("Menu snapshot", () => {
    const { asFragment } = render(
        <MediatorContext.Provider value={mediator}>
            <ServerContext.Provider value={server}>
                <Menu epages={() => {}} />
            </ServerContext.Provider>
        </MediatorContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
});
