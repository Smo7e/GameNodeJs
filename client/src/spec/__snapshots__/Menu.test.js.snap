// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu snapshot 1`] = `
<DocumentFragment>
  <div
    class="mainMenu"
    id="test-mainMemu"
  >
    <div
      style="height: 200px; width: 200px; position: absolute; bottom: 200px; right: 200px;"
    />
    <img
      class="photo-button"
      id="test-logo"
      src="logo.png"
    />
    <div
      class="buttons-container"
    >
      <div
        class="button1"
        id="test-play"
      >
        Играть
      </div>
      <div
        class="button2"
        id="test-heroes"
      >
        Герои
      </div>
      <div
        class="button3"
        id="test-settings"
      >
        Параметры
      </div>
    </div>
    <div
      class="profile-panel"
      id="test-profile"
    >
      <div
        class="user-profile"
        id="test-user"
      />
      <hr
        class="hr-user-profile1"
        id="test-hr1"
      />
      <div
        class="text-button"
        id="test-friends"
      >
        Друзья
      </div>
      <div
        class="text-Invitation"
        id="test-friends"
      >
        Приглашение
        <hr
          class="hr-line-Invitation"
          id="test-friends"
        />
      </div>
      <div
        class="new-profile-button"
      />
      <div
        style="left: 5px; height: 240px; width: 222px; position: absolute; bottom: 60px; right: 200px; overflow-y: hidden;"
      />
      <div
        class="your-friend-menu"
        id="test-friend-menu"
      >
        <div
          class="your-friend"
        >
          0: У Вас нету друзей,пхахахахахха!
        </div>
      </div>
      <hr
        class="hr-user-profile2"
        id="test-hr2"
      />
    </div>
    <button
      class="button-account"
      id="test-change-account"
    >
      Сменить аккаунт
    </button>
  </div>
</DocumentFragment>
`;
